src/main/java/com/teragrep/pth_06/planner/walker/conditions/ElementCondition.java

```diff
@@ -0,0 +1,134 @@
+        }
+        if (!element.hasAttribute("operation")) {
+            throw new IllegalStateException("Could not find specified or default value for 'operation' attribute from Element");
+        }
+        if (!element.hasAttribute("value")) {
+            throw new IllegalStateException("Could not find specified or default value for 'value' attribute from Element");
+        }
+    }
+
+    public ElementCondition(Element element, ConditionConfig config) {
+        this.element = element;
+        this.config = config;
+    }
+
+    public Condition condition() {
+        validate(element);
+        final String tag = element.getTagName();
+        final String operation = element.getAttribute("operation");
+        Condition condition = DSL.noCondition();
+        switch (tag.toLowerCase()) {
+            case "index":
+                QueryCondition index = new IndexCondition(element, config.streamQuery());
+                condition = index.condition();
+                break;
+            case "sourcetype":
+                QueryCondition sourceType = new SourceTypeCondition(element, config.streamQuery());
+                condition = sourceType.condition();
+                break;
+            case "host":
+                QueryCondition host = new HostCondition(element, config.streamQuery());
+                condition = host.condition();
+                break;
+        }
+        if (!config.streamQuery()) {
+            // Handle also time qualifiers
+            if ("earliest".equalsIgnoreCase(tag) || "index_earliest".equalsIgnoreCase(tag)) {
+                QueryCondition earliest = new EarliestCondition(element);
+                condition = earliest.condition();
+            }
+            if ("latest".equalsIgnoreCase(tag) || "index_latest".equalsIgnoreCase(tag)) {
+                QueryCondition latest = new LatestCondition(element);
+                condition = latest.condition();
+            }
+            // value search
+            if ("indexstatement".equalsIgnoreCase(tag) && "EQUALS".equals(operation) && config.bloomEnabled()) {
+                IndexStatementCondition indexStatementCondition =
+                        new IndexStatementCondition(element, config, new Tokenizer(32));
+                condition = indexStatementCondition.condition();
+            }
+        }
+        if (condition.equals(DSL.noCondition())) {
+            throw new IllegalStateException("Unsupported Element tag " + tag);
+        }
+        LOGGER.debug("Query condition: <{}>", condition);
+        return condition;
+    }
+
+    @Override
+    public boolean equals(final Object object) {
+        if (this == object) return true;
+        if (object == null) return false;
+        if (object.getClass() != this.getClass()) return false;
+        final ElementCondition cast = (ElementCondition) object;
+        return this.condition().toString().equals(cast.condition().toString());
```

Comment:
compare uses .toString() however toString is not implemented, it must use the object properties that make it equal to another i.e. in this case element and config

Reply:
refactored equals to compare only object properties